llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2181,6 +2181,120 @@ SmallVector<Range> UnPackOp::getIterationDomain(OpBuilder &builder) {
   return ::getIterationDomain(*this, builder);
 }
 
+SmallVector<Operation *>
+UnPackOp::getTiledImplementation(OpBuilder &builder,
+                                 ArrayRef<OpFoldResult> offsets,
+                                 ArrayRef<OpFoldResult> sizes) {
+  // TODO(hanchung): Extend it to handle memref version.
+  // Tiling on buffers needs extra buffer because tiled unpack op could produce
+  // more data for incomplete tiles. Tiling on tensors satisfies IREE's needs.
+  if (!hasTensorSemantics())
+    return {};
+
+  Location loc = getLoc();
+  auto ctx = builder.getContext();
+
+  AffineExpr dim0, dim1;
+  bindDims(ctx, dim0, dim1);
+  auto addMap = AffineMap::get(2, 0, {dim0 + dim1});
+  auto add = [&](OpFoldResult v1, OpFoldResult v2) -> OpFoldResult {
+    return makeComposedFoldedAffineApply(builder, loc, addMap, {v1, v2});
+  };
+  auto subMap = AffineMap::get(2, 0, {dim0 - dim1});
+  auto sub = [&](OpFoldResult v1, OpFoldResult v2) -> OpFoldResult {
+    return makeComposedFoldedAffineApply(builder, loc, subMap, {v1, v2});
+  };
+
+  int64_t inputRank = getInputRank();
+  int64_t outputRank = getOutputRank();
+  Attribute zeroAttr = builder.getIndexAttr(0);
+  Attribute oneAttr = builder.getIndexAttr(1);
+  DenseMap<int64_t, OpFoldResult> dimAndTileMapping = getDimAndTileMapping();
+  SmallVector<OpFoldResult> inputIndices, inputSizes, outputNewOffsets,
+      outputExpandedSizes;
+  for (auto dim : llvm::seq<int64_t>(0, outputRank)) {
+    if (dimAndTileMapping.count(dim)) {
+      DivModValue firstCoord =
```

Comment:
IIUC this is going to create a `affine.apply` operation. Maybe its worth just building the affine expr for div/mod here and then applying `makeComposedFoldedAffineApply` directly... that might canonicalize a lot better. Also avoids materializing the index ops for the `OpFoldResult` and the `affine.apply`.
FYI : Materializing `ConstantIndexOp`s can sometime lead to infinite loops (and hitting the loop limit in the pattern application) if the pattern application fails. You will end up creating index ops, and deleting them again and again.

Reply:
I gave it a shot, but the output IRs are identical. https://gist.githubusercontent.com/hanhanW/06c6a72b450546b03471802d8090ef27/raw We'll have to create a affine.apply op in either case. Because `offsets[i]` is definitely a Value type. It is an iv from `scf.for` loop. The affine maps are canonicalized well. Because the operands are all computed by makeComposedFoldedAffineApply methods.

In this case, LHS is always Value type, and RHS is always an attribute.

The firstCoord is canonicalized well because it's simple.

The lastCoord is also canonicalized well because the `add` and `sub` method use makeComposedFoldedAffineApply method.

In this context, I think it makes sense to keep the original code. What do you think?

Reply:
Sounds good.